First draft of a "common criteria" like standard for Cardano

RSoulatIOHK opened the pull request:
# Common Criteria

Common criteria is an international standard for cybersecurity. It is the merger of several national standards. It aims at international recognition of certificate issued by other countries. Most of the IT product that have security components in them have some level of Common Criteria certification.

I believe this could be a good starting point to establish a security standard for Cardano.

Common Criteria documentation is quite large.
My goal in this introduction is to only extract, at the moment, the main concepts and ideas from the introduction document to Common Criteria. This is not finished at all, I will try to have a "complete" introduction and would like to work on an example to see what works, what doesn't, what's missing etc.

All comments are more than welcomed!

RSoulatIOHK committed:
Create first draft

This is not finished at all. All comments are welcomed

RSoulatIOHK committed:
Removed the HTML part of MD

Wasn't being displayed on Github so I changed them

RSoulatIOHK committed:
updated the part on the ST

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+
+The Cardano Certification Standard (CCS) is a framework designed to facilitate the comparability of independent evaluations within the Cardano blockchain ecosystem. This standard establishes a comprehensive set of requirements for assessing the security features of Decentralized Applications (DApps) on the Cardano blockchain, as well as the assurance measures applied during security evaluations.
+
+The primary objective of the evaluation process is to establish a level of confidence that the security functionalities of these DApps and the assurance measures align with the predefined requirements. Consequently, the evaluation results serve as a valuable resource for end-users seeking to gauge whether a particular DApp aligns with their security requirements and expectations.
```
"align with the requirements defined here" otherwise it might be confused with particular requirements for particular DApps

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+
+The Cardano Certification Standard also serves as a guide for both DApp developers and evaluators. It offers insights and best practices for the development of DApps with robust security features and provides a structured approach for evaluating their security posture.
+
+Flexibility is a core principle of the Cardano Certification Standard. It is designed to accommodate a diverse range of evaluation methodologies, ensuring that all essential security properties required by a DApp can be effectively assessed. However, it is essential to note that the validity of an evaluation is contingent upon the specific security properties evaluated and the methods employed. A ...
```
"to assess the significance of the evaluation results"

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+
+Flexibility is a core principle of the Cardano Certification Standard. It is designed to accommodate a diverse range of evaluation methodologies, ensuring that all essential security properties required by a DApp can be effectively assessed. However, it is essential to note that the validity of an evaluation is contingent upon the specific security properties evaluated and the methods employed. A ...
+
+The Cardano Certification Standard aims at adressing:
```
"aims to address:"

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+The Cardano Certification Standard aims at adressing:
+
+- The protection of assets of any user interacting with the DApp
+- The loss of use of the DApp
```
I was trying to reformulate this. It's something like "Ensuring the accessibility and liveness of the  services that the DApp provides."

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+
+In conclusion, the Cardano Certification Standard serves as a document promoting transparency, consistency, and confidence in the security of DApps.
+
+*RSoulat* This document draws heavily from the work of Common Criteria. I think we should take some inspiration from it for the security aspects, and, if we want to deal with purely functional aspects, add them where we see fit.
```
General comment that I may as well make here.

Given the dependency on the common criteria, I wondered whether that should be more explicit. That could be done in two ways.

1. By including cross references to the common criteria standards in appropriate places in this document.
2. By writing a different kind of document that is effectively a diff with the common criteria document.

The point about doing this is that it makes comparison with the existing standard much easier, and also allows evolution of this standard as it grows and also evolution with the common criteria standard as it evolves.

RSoulatIOHK replied:
My issue with doing a diff is that the CC are a very large set of documents for the security of IT products. Some of the concepts, I believe don't need to be applied to our situation and I'm not sure it would make sense to have a diff. 
I agree we could cross-reference though.

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+
+### 2.1 Scope
+
+The CCS wishes to establish the general concepts and principles of DApp security evaluation and to specify the general model of evauation. This section provides an overview of all the parts of the CCS, it defines all the different terms, abbreviations and concepts used through the rest of the document. It establishes the core principles of the standard:
```
evaluation

simonjohnthompson commented on Certification-Standard/introduction.md:
```diff
@@ -0,0 +1,239 @@
+### 2.2 Terms and Definitions
+
+*RSoulat* Note: Taken from the CC definitions
+*RSoulat* Maybe we could drop the Protection Profiles from a first version and only deal with Security Target
```
I'd encourage taking a minimalist approach: only include things if it's really necessary. In particular, something like profiles might just be a reference back to the common criteria standard for information, for example.

RSoulatIOHK replied:
I agree. 

The Protection Profiles are interesting but I think they will come later in the use of the standard. For example, a DEX might come in the future with a more standardized set of security requirements that needs to be verified and this could become a Profile.

simonjohnthompson commented on Certification-Standard/introduction.md:
Set of logical interfaces to a target, which provide access to the target's assets or control.

simonjohnthompson commented on Certification-Standard/introduction.md:
Potentially accessing (third-party) services too. For example, blockfrost.

simonjohnthompson commented on Certification-Standard/introduction.md:
Maybe the general notion of "digital asset" would be useful here.

simonjohnthompson commented on Certification-Standard/introduction.md:
We should explain why there's this divergence here: I guess the reason is (im)maturity.

RSoulatIOHK replied:
Sorry it's not very clear in the text. In CC, the evaluator is not the certificate provider. Usually the certificate provider is a state entity like ANSSI for France, NIAP for the US etc. It's because the certificate provider also has some verification to do. The evaluator has some activity it needs to perform, documents and reports that need to be produced etc. So there is a need for a third party to check that the evaluator has also done its work according to the CC.

simonjohnthompson commented on Certification-Standard/introduction.md:
Is there any reason to be so specific?

RSoulatIOHK replied:
Maybe not, after reading it once again, UTxO probably falls under the definition

simonjohnthompson commented on Certification-Standard/introduction.md:
Environment

simonjohnthompson commented on Certification-Standard/introduction.md:
… but it some cases the TOE will be wider, including off-chain code and services.

RSoulatIOHK replied:
Absolutely, I was just giving some examples. Third-party services will probably always fall under OE though.

simonjohnthompson commented on Certification-Standard/introduction.md:
Does the common criteria standard have a clear distinction between concepts of vulnerability and weakness? We should be consistent with that.

RSoulatIOHK replied:
There is no definition given to weakness in the CC but the meaning is close to "potential vulnerability".
The only difference is that a Weakness is more general than a "potential vulnerability". A weakness can be "Integer overflow", a potential vulnerability would be "on line 33, there is a risk of integer overflow, but we are not able to actually reach it".

simonjohnthompson commented on Certification-Standard/introduction.md:
There is anecdotal evidence that there is a correlation between good development practices and better security properties of DApps.

RSoulatIOHK replied:
Absolutely. But I don't think we can include that as a "security objective" it's actually part of the ADV activities in CC.

simonjohnthompson commented on Certification-Standard/introduction.md:
Software and services?

RSoulatIOHK replied:
What would be a service that is not software?

simonjohnthompson commented on Certification-Standard/introduction.md:
"users implementation-independent"

simonjohnthompson commented on Certification-Standard/introduction.md:
independent

RSoulatIOHK replied:
That's my French taking over

simonjohnthompson commented on Certification-Standard/introduction.md:
libraries and services

simonjohnthompson commented on Certification-Standard/introduction.md:
misrepresented

simonjohnthompson commented on Certification-Standard/introduction.md:
refer

simonjohnthompson commented on Certification-Standard/introduction.md:
addressed

simonjohnthompson commented on Certification-Standard/introduction.md:
what's the purpose of "functional and assurance" here?

RSoulatIOHK replied:
Roughly, the functional requirements are what the TOE should do. The assurance requirements are how it should be tested, what activities should be done to ensure that it is compliant.

For example, the assurance requirements for a DApp could say: I want my TOE to be evaluated for a Level 3 certification. This would imply some activities for the developer had to perform some kind activities in the definition of  the formal model for example and provide some results/documentation (like proofs, assumptions, solver used etc.)

RSoulatIOHK replied:
But it might be too much for the moment, and maybe we can focus on functional requirements for now

simonjohnthompson commented on Certification-Standard/introduction.md:
agreed

simonjohnthompson commented on Certification-Standard/introduction.md:
Are evaluators not involved at this level at all?

RSoulatIOHK replied:
That's an open question I believe. Do we require that no matter what, an evaluator has to do the certification or should an automated system be able to do it?

simonjohnthompson commented on Certification-Standard/introduction.md:
free of vulnerabilities? how about "without identified vulnerabilities"?

simonjohnthompson commented on Certification-Standard/introduction.md:
Not sure about "full" here. Must be formal verification of an identified set of security properties relative to an explicitly articulated set of assumptions.

RSoulatIOHK replied:
I think we agree but "full" is probably not the right term. I meant full in the sense that everything inside the TOE that should be verified has been verified given the assumptions about the OE.

RSoulatIOHK committed:
Update following Simon's comments

RSoulatIOHK committed:
Moved the old document to "draft" and rewrote a lighter suggestion for  certification standard.

It still lacks complance to CIP-0052, reference to the certification level and where they need to be taken into account.
The text is still too "test" oriented and doesn't easily leave room for formal verification, maybe it can be changed in the text with "verified" instead of "tested".
We'll be able to loop back to certification levels  when we better describe the things the developer (or auditor, or sponsor if we want to add them?) needs to deliver for the evaluation. We could have layers for spec for example, level 3 requiring a formal spec and a formal verification done on those spec and/or on the conformance of source code to formal spec.

aleeusgr commented:
Cardano Well Architected Framework (CWAF)
describes key concepts, design principles, and architectural best practices for designing and using ...

https://github.com/input-output-hk/Developer-Experience-working-group/issues/28

aleeusgr commented:
https://medium.com/@alexeusgr/adopting-the-common-criteria-in-a-decentralized-organisation-challenges-and-opportunities-4b438d489717

RSoulatIOHK commented:
> Developing a Well Architected Framework alongside the Certification Standard is an opportunity to improve projects health through increased visibility and awareness, collaboration opportunities, personal development for contributors and strengthening the community.
> 
> https://medium.com/@alexeusgr/adopting-the-common-criteria-in-a-decentralized-organisation-challenges-and-opportunities-4b438d489717

Thanks for the link! It looks interesting, I'll read through it in more details
